Add support for the new f-string tokens per PEP 701

dhruvmanila opened the pull request:
## Summary

This PR adds support in the lexer for the newly added f-string tokens as per PEP 701. The following new tokens are added:
* `FStringStart`: Token value for the start of an f-string. This includes the `f`/`F`/`fr` prefix and the opening quote(s).
* `FStringMiddle`: Token value that includes the portion of text inside the f-string that's not part of the expression part and isn't an opening or closing brace.
* `FStringEnd`: Token value for the end of an f-string. This includes the closing quote.

Additionally, a new `Exclamation` token is added for conversion (`f"{foo!s}"`) as that's part of an expression.

## Test Plan

New test cases are added to for various possibilities using snapshot testing. The output has been verified using python/cpython@f2cc00527e.

## Benchmarks

_I've put the number of f-strings for each of the following files after the file name_

```
lexer/large/dataset.py (1)       1.05   612.6±91.60µs    66.4 MB/sec    1.00   584.7±33.72µs    69.6 MB/sec
lexer/numpy/ctypeslib.py (0)     1.01    131.8±3.31µs   126.3 MB/sec    1.00    130.9±5.37µs   127.2 MB/sec
lexer/numpy/globals.py (1)       1.02     13.2±0.43µs   222.7 MB/sec    1.00     13.0±0.41µs   226.8 MB/sec
lexer/pydantic/types.py (8)      1.13   285.0±11.72µs    89.5 MB/sec    1.00   252.9±10.13µs   100.8 MB/sec
lexer/unicode/pypinyin.py (0)    1.03     32.9±1.92µs   127.5 MB/sec    1.00     31.8±1.25µs   132.0 MB/sec
```

It seems that overall the lexer has regressed. I profiled every file mentioned above and I saw one improvement which is done in (098ee5d493ca83238754a8cb4629fa1b91144b84). But otherwise I don't see anything else. A few notes by isolating the f-string part in the profile:
* As we're adding new tokens and functionality to emit them, I expect the lexer to take more time because of more code.
* The `lex_fstring_middle_or_end` takes the most amount of time followed by the `current_mut` line when lexing the `:` token. The latter is to check if we're at the start of a format spec or not.
* In a f-string heavy file such as https://github.com/python/cpython/blob/main/Lib/test/test_fstring.py [^1] (293), most of the time in `lex_fstring_middle_or_end` is accounted by string allocation for the string literal part of `FStringMiddle` token (https://share.firefox.dev/3ErEa1W)

I don't see anything out of ordinary for `pydantic/types` profile (https://share.firefox.dev/45XcLRq)

fixes: #7042

[^1]: We could add this in lexer and parser benchmark

dhruvmanila commented:
Current dependencies on/for this PR:
* main
  * **PR #6658** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6658" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #6659** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6659" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/6659?utm_source=stack-comment).

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -51,6 +51,9 @@ use crate::{
 mod cursor;
 mod indentation;
 
+/// Maximum nesting level of f-string expressions.
+const MAX_FSTRING_EXPRESSION_NESTING: u32 = 3;
```
Is there a technical reason why we want to limit the fstring nesting?

dhruvmanila replied:
No, PEP says that the lower bound is 2. While there's no upper bound, CPython implementation itself have an [upper bound of 3](https://github.com/python/cpython/blob/fd195092204aa7fc9f13c5c6d423bc723d0b3520/Parser/tokenizer.h#L43) and [here's](https://github.com/python/cpython/blob/fd195092204aa7fc9f13c5c6d423bc723d0b3520/Parser/tokenizer.c#L2650-L2652) where they throw an error if it goes beyond that. Currently, we have an upper bound of 2

https://github.com/astral-sh/ruff/blob/0cea4975fcfe09716c084c0a18d1b4c4cd9b8f05/crates/ruff_python_parser/src/string.rs#L420-L422

MichaReiser replied:
I would remove the bound because different interpreter implementations may use different limits. However, it would be a good idea to have a lint rule that warns about too deeply nested FString depending on your target runtime.

dhruvmanila replied:
There's also the limitation of nested f-strings itself, so `f"foo {f"bar {f"baz"}"}"` which is also [limited to 150](https://github.com/python/cpython/blob/0cb0c238d520a8718e313b52cffc356a5a7561bf/Parser/tokenizer.h#L15). I'm avoiding that in our implementation for the same reason.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +514,138 @@ impl<'source> Lexer<'source> {
+    fn try_lex_fstring_middle(&mut self) -> Result<Option<Tok>, LexicalError> {
+        let mut in_named_unicode = false;
+
+        // SAFETY: Safe because the function is only called when `self.fstring_stack` is not empty.
+        let context = self.fstring_stack.last().unwrap();
```
Could we pass the fstring context instead of using an `unwrap` here? It isn't just safer but also guarantees that we don't have a panic branch in release builds.

dhruvmanila replied:
Not really. The `self.try_lex_fstring` functions borrow `self` as mutable but then we pass in an immutable borrow of `FStringContext` which belongs to `self`:

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> crates/ruff_python_parser/src/lexer.rs:730:36
    |
727 |         if let Some(fstring_context) = self.fstring_stack.last() {
    |                                        ------------------------- immutable borrow occurs here
...
730 |                 if let Some(tok) = self.try_lex_fstring_middle(fstring_context)? {
    |                                    ^^^^^----------------------^^^^^^^^^^^^^^^^^
    |                                    |    |
    |                                    |    immutable borrow later used by call
    |                                    mutable borrow occurs here
```

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +514,138 @@ impl<'source> Lexer<'source> {
+                        return Some(Tok::FStringEnd);
+                    }
+                    StringQuoteSize::Triple => {
+                        let mut remaining = self.cursor.rest().chars().clone();
```
Why do we need to clone chars?

dhruvmanila replied:
Oh right, we don't.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +514,138 @@ impl<'source> Lexer<'source> {
+        // SAFETY: Safe because the function is only called when `self.fstring_stack` is not empty.
+        let context = self.fstring_stack.last().unwrap();
+
+        if let quote @ ('\'' | '"') = self.cursor.first() {
+            if quote == context.quote_char() {
```
```suggestion
        if self.cursor.first() == context.quote_char() {
```

dhruvmanila replied:
We actually need the `quote` for the lookaheads if it's a triple quoted string.

MichaReiser replied:
Could we use `context.quote_char`?

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1292,108 @@ impl std::fmt::Display for LexicalErrorType {
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum StringQuoteSize {
```
Nit: Maybe `Kind`

dhruvmanila replied:
There's a similar implementation for the string quote and size in the `tokenizer.rs` file where `SimpleTokenizer` is present. But, we can't really depend on `ruff_python_trivia`. Currently, I've added an additional implementation here but maybe it could be possible to move it to a common place like `ruff_python_ast`?

dhruvmanila commented:
@MichaReiser Thanks for your initial review even though the PR is still in draft :)

konstin commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -551,17 +551,20 @@ impl<'source> Lexer<'source> {
-                            self.cursor.bump();
+                            self.cursor.bump(); // 'N'
+                            self.cursor.bump(); // '{'
                             in_named_unicode = true;
```
Do we want to add `eat_if` and `eat_if2` methods?

dhruvmanila replied:
Actually, we could use `eat_char` in the same order as it will short circuit if the first char isn't `N`.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -523,6 +658,10 @@ impl<'source> Lexer<'source> {
             false
         };
 
+        if kind.is_any_fstring() {
+            return Ok(self.lex_fstring_start(quote, triple_quoted, kind.is_raw()));
+        }
+
```
I'm thinking of removing the `StringKind::FString` and `StringKind::RawFString` as that is an invalid representation now that we'll be emitting different tokens for f-strings. Instead, I'm thinking of updating the `lex_identifier` to directly check for `f` (and related `F`, `r`, `R`) and call `lex_fstring_start` directly.

dhruvmanila replied:
This will be done after the linter changes are complete.

dhruvmanila commented:
(This is ready for review, but not to be merged)

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
+    /// Returns `true` if the colon (`:`) for the current f-string is in a valid
+    /// position i.e., at the same level of nesting as the opening parentheses token.
+    /// Increments the number of format specs if it is.
+    fn is_valid_format_spec_position(&mut self) -> bool {
```
Nit: It's rather unexpected that a `is_` function mutates self. Maybe `try_start_format_spec`

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
+    /// position i.e., at the same level of nesting as the opening parentheses token.
+    /// Increments the number of format specs if it is.
+    fn is_valid_format_spec_position(&mut self) -> bool {
+        if self.parentheses_count - self.format_spec_count == 1 {
```
This could potentially overflow, or is it guaranteed that `parentheses_count` is always larger than `format_spec_count` (even for invalid input)? 

Edit: This doesn't seem to be given, see `is_in_expression`. We need to use `saturating_sub` here

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
+struct FStringContext {
+    quote_char: StringQuoteChar,
+    quote_size: StringQuoteSize,
+    parentheses_count: u32,
+    format_spec_count: u32,
```
Can we document these fields and why tracking them in the state is necessary? E.g. why do we need to track both `parenthesis_count` and `format_spec_count` (Should it be `format_spec_depth`?). What are the different possible states?

dhruvmanila replied:
There are 3 reasons to track these fields:
1. To check if we're in a f-string expression `f"{<here>}"`
2. To check if we're in a format spec within a f-string expression `f"{foo:<here>}"`
3. To check if the `:` is to start a format spec or is it part of, for example, named expression `:=`:
	* For `f"{x:=1}"`, the colon is to indicate the format spec start
	* For `f"{(x:=1)}"`, the colon is part of `:=` named expression because the colon is not at the same level of parentheses. Another example is `f"{x,{y:=1}}"` where the colon is part of `:=` named expression

From [PEP 701: How to produce these new tokens](https://peps.python.org/pep-0701/#how-to-produce-these-new-tokens):
> 3. [..] This mode tokenizes as the “Regular Python tokenization” until a `:` or a `}` character is encountered with the same level of nesting as the opening bracket token that was pushed when we enter the f-string part. [..]

With this explanation, do you think `_depth` is a better suffix?

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
     }
 }
 
+#[derive(Debug)]
+enum StringQuoteChar {
+    Single,
+    Double,
+}
```
Nit: Implement `Copy` for these value enums. Rust will then tell you that it's more performant to pass `StringQuoteChar` by value rather by reference (1 byte vs 8 byte + dereferencing)

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -847,39 +999,66 @@ impl<'source> Lexer<'source> {
             }
             '~' => Tok::Tilde,
             '(' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.open_parentheses();
+                }
                 self.nesting += 1;
```
It is unfortunate that we now need to increment open parentheses twice if inside of an `fstring` context (and we have more branching inside a hot function. 

Would it be possible to store the initial `parentheses_count` on the `FStringContext` instead? You can then compute the fstring parentheses by doing `self.nesting.saturating_sub(self.initial_count)`

dhruvmanila replied:
This is a good point! This would mean that to get the `parentheses_count` inside any method of `FStringContext`, the `self.nesting` needs to be passed to these methods.

Also, note that for closing parentheses we still need to call the method to decrement the format spec count.

dhruvmanila replied:
I've made this change. It turns out that the `last_mut` part was taking on average 10% of the lexing time. I think this is because we were performing this check for _every_ parentheses irrespective of whether we're inside a f-string or not. This has improved the benchmarks. Thanks!

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -157,6 +159,7 @@ impl<'source> Lexer<'source> {
             source: input,
             cursor: Cursor::new(input),
             mode,
+            fstring_stack: vec![],
```
Nit: Have you considered introducing a `FStringStack` newtype wrapper around `Vec`? Could that reduce some repetitive code inside of the lexer? Similar to `Indentations`

dhruvmanila replied:
We can introduce that abstraction but can you expand on "Could that reduce some repetitive code inside of the lexer?". What part is repetitive?

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -847,39 +999,66 @@ impl<'source> Lexer<'source> {
             }
             ':' => {
-                if self.cursor.eat_char('=') {
+                if self
+                    .fstring_stack
+                    .last_mut()
+                    .is_some_and(FStringContext::is_valid_format_spec_position)
+                {
+                    Tok::Colon
+                } else if self.cursor.eat_char('=') {
```
Is this because `:=` has a different meaning inside of a format spec?

dhruvmanila replied:
Does https://github.com/astral-sh/ruff/pull/6659#discussion_r1303742193 answer your question here?

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
+    quote_size: StringQuoteSize,
+    parentheses_count: u32,
+    format_spec_count: u32,
+    is_raw_string: bool,
```
Nit: We could potentially squeeze `quote_char`, `quote_size` and `raw` (3 bytes) into a single byte by using bitflags. It won't give us any space-saving because of padding but may be easier to expand.

You would have three flags:

* TRIPLE (missing implies single)
* DOUBLE (missing implies single)
* RAW (missing implies non-raw)

dhruvmanila replied:
This is a good idea! I've implemented it and it makes the code which uses the `FStringContext` better to reason about.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -579,6 +718,19 @@ impl<'source> Lexer<'source> {
+                    self.fstring_stack.pop();
+                    return Ok((tok, self.token_range()));
+                }
+            }
```
Nit: I wonder if it would make sense to refactor `next_token` a bit so that we could call into a `lex_fstring_expression` that:

* Doesn't handle indentation
* overrides the lexing for fstring- specific tokens 

But I'm not sure if its worth the complexity (and work)

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+                                break;
+                            }
+                        }
+                    }
+                    self.cursor.bump();
+                }
+                '{' => {
+                    if self.cursor.second() == '{' {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                '}' => {
+                    if in_named_unicode {
+                        in_named_unicode = false;
+                        self.cursor.bump();
+                    } else if self.cursor.second() == '}' && !context.is_in_format_spec() {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                _ => {
+                    self.cursor.bump();
+                }
+            }
+        }
+        let value_end = self.offset();
+
+        if value_start == value_end {
+            return Ok(None);
+        }
+
+        let value = self.source[TextRange::new(value_start, value_end)]
+            .to_string()
+            .replace("{{", "{")
+            .replace("}}", "}");
```
Nit
```suggestion
        self.cursor.start_token();
        loop {
            match self.cursor.first() {
                EOF_CHAR => {
                    let error = if context.allow_multiline() {
                        FStringErrorType::UnterminatedTripleQuotedString
                    } else {
                        FStringErrorType::UnterminatedString
                    };
                    self.fstring_stack.pop();
                    return Err(LexicalError {
                        error: LexicalErrorType::FStringError(error),
                        location: self.offset(),
                    });
                }
                '\n' if !context.allow_multiline() => {
                    self.fstring_stack.pop();
                    return Err(LexicalError {
                        error: LexicalErrorType::FStringError(FStringErrorType::UnterminatedString),
                        location: self.offset(),
                    });
                }
                '\\' => {
                    self.cursor.bump(); // '\'
                    if matches!(self.cursor.first(), '{' | '}') {
                        // Don't consume `{` or `}` as we want them to be consumed as tokens.
                        break;
                    } else if !context.is_raw_string {
                        if self.cursor.eat_char('N') && self.cursor.eat_char('{') {
                            in_named_unicode = true;
                            continue;
                        }
                    }
                    // Consume the escaped character.
                    self.cursor.bump();
                }
                quote @ ('\'' | '"') if quote == context.quote_char() => {
                    match context.quote_size {
                        StringQuoteSize::Single => break,
                        StringQuoteSize::Triple => {
                            let mut remaining = self.cursor.rest().chars().skip(1);
                            if remaining.next() == Some(quote) && remaining.next() == Some(quote) {
                                break;
                            }
                        }
                    }
                    self.cursor.bump();
                }
                '{' => {
                    if self.cursor.second() == '{' {
                        self.cursor.bump();
                        self.cursor.bump();
                    } else {
                        break;
                    }
                }
                '}' => {
                    if in_named_unicode {
                        in_named_unicode = false;
                        self.cursor.bump();
                    } else if self.cursor.second() == '}' && !context.is_in_format_spec() {
                        self.cursor.bump();
                        self.cursor.bump();
                    } else {
                        break;
                    }
                }
                _ => {
                    self.cursor.bump();
                }
            }
        }
        let range = self.token_range():

        if range.is_empty() {
            return Ok(None);
        }

        let value = self.source[range]
            .to_string()
            .replace("{{", "{")
            .replace("}}", "}");
```

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+        }
+
+        let value = self.source[TextRange::new(value_start, value_end)]
+            .to_string()
```
The `to_string` call here is unnecessary. `replace` always allocates a new string already
```suggestion
```

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+
+        let value = self.source[TextRange::new(value_start, value_end)]
+            .to_string()
+            .replace("{{", "{")
+            .replace("}}", "}");
```
Calling replace twice is `O(2N)` on the string length. This may not matter much considering that most strings are small but can add up. It is also unfortunate that we do this even for strings that we know never contain a `{{` or `}}` sequence. There are two ways of how we can avoid this additional pass over the string:

* Track whether the string contained any `{{` or `}}` and only then call replace
* Implement your own `replace` that replaces both sequences in a single loop 
* Create a normalized string, but only initialize it when you see a `{{` or `}}`. Otherwise use the full text range, see https://github.com/astral-sh/ruff/blob/0cea4975fcfe09716c084c0a18d1b4c4cd9b8f05/crates/ruff_python_formatter/src/expression/string.rs#L607-L679

dhruvmanila replied:
Thanks! I've implemented this in [`2de590d` (#6659)](https://github.com/astral-sh/ruff/pull/6659/commits/2de590d2341dfd7f2f5b86458dc1bcba097ab99d)

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+        Ok(Some(Tok::FStringMiddle(value)))
+    }
+
+    fn maybe_lex_fstring_end(&mut self) -> Option<Tok> {
```
```suggestion
    fn eat_fstring_end(&mut self) -> Option<Tok> {
```

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+        // SAFETY: Safe because the function is only called when `self.fstring_stack` is not empty.
+        let context = self.fstring_stack.last().unwrap();
+
+        if self.cursor.first() == context.quote_char() {
+            match context.quote_size {
+                StringQuoteSize::Single => {
+                    self.cursor.bump();
+                    return Some(Tok::FStringEnd);
+                }
+                StringQuoteSize::Triple => {
+                    let mut remaining = self.cursor.rest().chars().skip(1);
+                    if remaining.next() == Some(context.quote_char())
+                        && remaining.next() == Some(context.quote_char())
+                    {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                        self.cursor.bump();
+                        return Some(Tok::FStringEnd);
+                    }
+                }
+            }
+        }
```
I prefer to use `eat` whenever possible because it avoids bugs where I accidentally forgot to call `bump`, which results in an infinite loop:

```suggestion
        match context.quote_size {
            StringQuoteSize::Single => {
                if self.cursor.eat_char(context.quote_char()) {
                    return Some(Tok::FStringEnd);
                }
            }
            StringQuoteSize::Triple => {
                let expected = match context.quote_char {
                    StringQuoteChar::Single => "'''",
                    StringQuoteChar::Double => r#"""""#,
                };
                if self.cursor.rest().starts_with(expected) {
                    self.cursor.bump();
                    self.cursor.bump();
                    self.cursor.bump();
                    return Some(Tok::FStringEnd);
                }
            }
        }
```

I like @konstin's suggestion to add a `eat_char2` and `eat_char3` that only eat if all two or three chars match.

dhruvmanila replied:
Oh yes, I can totally relate to the part where we forgot to call `bump` :)

I've added these methods.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -579,6 +718,19 @@ impl<'source> Lexer<'source> {
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            if !fstring_context.is_in_expression() {
+                self.cursor.start_token();
+                if let Some(tok) = self.maybe_lex_fstring_middle()? {
+                    return Ok((tok, self.token_range()));
+                }
+                if let Some(tok) = self.maybe_lex_fstring_end() {
+                    self.fstring_stack.pop();
+                    return Ok((tok, self.token_range()));
+                }
```
For which tokens does fstring_middle` and `fstring_end` both return None? Could we instead branch early on these tokens and unconditionally either return the middle or end part?

dhruvmanila replied:
So, `maybe_lex_fstring_middle` will return `None` when:
1. The first character is `{` not followed by another `{` which is the start of f-string expression. For example, `f"{foo}"` and not `f"{{foo}}"`
2. Empty f-string which means that we should directly consume the f-string end. For example, `f""`, `f''`, `f""""""`, `f''''''`.

I think you're referring to (1) then.

dhruvmanila replied:
This is a really good suggestion. Thanks, I've added it here [`5b41241` (#6659)](https://github.com/astral-sh/ruff/pull/6659/commits/5b41241b0f37d7539439c93a3408d91f69c29861) with some test cases.

dhruvmanila replied:
I've added `unreachable` to check in what other cases this happens.

dhruvmanila replied:
There's another case as well:

We're in format specifier mode and we've reached the end i.e., the next character is `}` which should be emitted as RBrace.

For example, in `f"{foo:.3f}"`, the format specifier `.3f` is emitted as `FStringMiddle`, then in the next iteration we need to emit `}` (RBrace) but we'll try to lex f-string middle (returns `None`) and we're not at the end of f-string.

Now, I'm unable to find any other case which reaches `unreachable`.

dhruvmanila replied:
I do like this idea but it sacrifices readability a bit (https://github.com/astral-sh/ruff/pull/6659#discussion_r1305569804). I'll give it a bit of thought but I'm mostly leaning towards reverting the change.

MichaReiser commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -579,6 +718,19 @@ impl<'source> Lexer<'source> {
     // This is the main entry point. Call this function to retrieve the next token.
     // This function is used by the iterator implementation.
     pub fn next_token(&mut self) -> LexResult {
+        if let Some(fstring_context) = self.fstring_stack.last() {
```
I dislike adding a stack lookup into the hot path but I don't see a way of how we could avoid it :(

dhruvmanila replied:
Maybe we could utilize `State`? The problem here is that it'll get updated when inside a f-string expression as it's not persistent.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -579,6 +728,28 @@ impl<'source> Lexer<'source> {
     // This function is used by the iterator implementation.
     pub fn next_token(&mut self) -> LexResult {
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            if !fstring_context.is_in_expression()
+                // Avoid lexing f-string middle/end if we're sure that this is
+                // the start of a f-string expression i.e., `f"{foo}"` and not
+                // `f"{{foo}}"`.
+                && (self.cursor.first() != '{' || self.cursor.second() == '{')
+                // Avoid lexing f-string middle/end if we're sure that this is
+                // the end of a f-string expression. This is only for when
+                // the `}` is after the format specifier i.e., `f"{foo:.3f}"`
+                // because the `.3f` is lexed as `FStringMiddle` and thus not
+                // in a f-string expression.
+                && (!fstring_context.is_in_format_spec() || self.cursor.first() != '}')
+            {
```
I don't really like this that much. I think we're sacrificing readability a bit. Without this, the `lex_fstring_middle_or_end` function would return `Result<Option<Tok>, ...>` and this condition will be identified initially in the function loop itself.

dhruvmanila replied:
I've reverted this change although I'm open to suggestions.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -168,6 +173,34 @@ impl<'source> Lexer<'source> {
         // Detect potential string like rb'' b'' f'' u'' r''
+        #[cfg(feature = "pep-701")]
+        match (first, self.cursor.first()) {
+            ('f' | 'F', quote @ ('\'' | '"')) => {
+                self.cursor.bump();
+                return Ok(self.lex_fstring_start(quote, false));
+            }
+            ('r' | 'R', 'f' | 'F') | ('f' | 'F', 'r' | 'R') if is_quote(self.cursor.second()) => {
+                self.cursor.bump();
+                let quote = self.cursor.bump().unwrap();
+                return Ok(self.lex_fstring_start(quote, true));
+            }
```
This separation is because the `StringKind::FString` and `StringKind::RawFString` variants will be removed in the parser changes.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +541,161 @@ impl<'source> Lexer<'source> {
+    /// Lex a f-string start token.
+    #[cfg(feature = "pep-701")]
+    fn lex_fstring_start(&mut self, quote: char, is_raw_string: bool) -> Tok {
+        #[cfg(debug_assertions)]
+        debug_assert_eq!(self.cursor.previous(), quote);
+
+        // If the next two characters are also the quote character, then we have a triple-quoted
+        // string; consume those two characters and ensure that we require a triple-quote to close
+        let quote_size = if self.cursor.first() == quote && self.cursor.second() == quote {
+            self.cursor.bump();
+            self.cursor.bump();
+            StringQuoteSize::Triple
+        } else {
+            StringQuoteSize::Single
+        };
```
This is repetitive from the `lex_string` function, maybe we could define a `consume_open_quote` function:

```rust
fn consume_open_quote(&mut self, quote: char) -> (StringQuoteChar, StringQuoteSize) {
	// ...
}
```

Or, return bitflags?

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +541,161 @@ impl<'source> Lexer<'source> {
+        };
+        Ok(Some(Tok::FStringMiddle {
+            value,
+            is_raw: context.is_raw_string,
```
This is required for the parser changes.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -523,6 +711,30 @@ impl<'source> Lexer<'source> {
 
+        #[cfg(feature = "pep-701")]
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            // When we are in an f-string, check whether does the initial quote
+            // matches with f-strings quotes and if it is, then this must be a
+            // missing '}' token so raise the proper error.
```
Source: https://github.com/python/cpython/blob/21a7420190778fb6e9237bf12e029a26cd18d82d/Parser/tokenizer.c#L2444-L2455

dhruvmanila commented on crates/ruff_python_parser/src/parser.rs:
```diff
@@ -698,9 +699,9 @@ array[3:5, *indexes_to_select]
     raise ValueError(1)
 except TypeError as e:
-    print(f'caught {type(e)}')
+    print('caught', type(e))
 except OSError as e:
-    print(f'caught {type(e)}')"#,
+    print('caught', type(e))"#,
```
The tests are for `try ... except ...` so let's use normal strings to avoid putting the test behind feature flag.

dhruvmanila commented on crates/ruff_python_parser/src/parser.rs:
```diff
@@ -714,9 +715,9 @@ except OSError as e:
         [ValueError(1), TypeError(2), OSError(3), OSError(4)])
 except* TypeError as e:
-    print(f'caught {type(e)} with nested {e.exceptions}')
+    print('caught', type(e), 'with nested', e.exceptions)
 except* OSError as e:
-    print(f'caught {type(e)} with nested {e.exceptions}')"#,
+    print('caught', type(e), 'with nested', e.exceptions)"#,
```
The tests are for `try ... except ...` so let's use normal strings to avoid putting the test behind feature flag.

konstin commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -508,6 +521,148 @@ impl<'source> Lexer<'source> {
+            return Ok(Some(Tok::FStringEnd));
+        }
+
+        // The normalized string if the token value is not yet normalized.
+        // This must remain empty if it's already normalized. Normalization
+        // is to replace `{{` and `}}` with `{` and `}` respectively.
```
```suggestion
        // We have to decode `{{` and `}}` into `{` and `}` respectively. As an optimization, we 
        // only allocate a new string we find any escaped curly braces, otherwise this string will
        // remain empty and we'll use a source slice instead.
```

dhruvmanila commented:
## Major changes since last review:

* Combine lexing of `FStringMiddle` and `FStringEnd` tokens (earlier they were separate).
* Added `FStrings` (similar to `Indentations`), `FStringContextFlags` and related changes in the lexer.
* Emit empty `FStringMiddle` token for special case to signal the lexer that we're in a lambda expression without parentheses so that the parser throws an error instead.
* Avoid multiple increment/decrement for nesting. Instead store the initial nesting in `FStringContext` and pass in the current nesting to compute the f-string open parentheses on demand.

MichaReiser reviewed (approved):
I've a few more nit comments but this looks good to me. I'm not too worried about the performance regression because we move logic from the parser to the lexer, meaning these changes should ultimately lead to a perf improvement in the parser.

dhruvmanila commented on crates/ruff_python_parser/src/lexer.rs:
```diff
@@ -168,16 +173,24 @@ impl<'source> Lexer<'source> {
-            second @ ('f' | 'F' | 'r' | 'R' | 'b' | 'B') if is_quote(self.cursor.second()) => {
+            (_, second @ ('r' | 'R' | 'b' | 'B')) if is_quote(self.cursor.second()) => {
                 self.cursor.bump();
-
```
My reasoning behind this change is that the `FString` variant would be an invalid state as there won't be any `String` token of that kind. It just becomes an intermediate state to determine whether we want to lex a string or f-string start token. So, I was planning to remove it once all the linter changes are complete because they still use this variant.

dhruvmanila commented on crates/ruff_python_parser/src/lexer/fstring.rs:
```diff
@@ -0,0 +1,155 @@
+        self.flags.contains(FStringContextFlags::TRIPLE)
+    }
+
+    fn open_parentheses_count(&self, current_nesting: u32) -> u32 {
```
This actually includes all the three types of parentheses: `(`, `{` and `[`.

dhruvmanila commented:
***The implementation have been removed in #7263. Keeping this here for posterity:***

> ### Empty `FStringMiddle` token
>
> `FStringMiddle` includes the portion of text inside the f-string that's not part of the expression and isn't an opening or closing brace which also includes the format spec. For example, in `f"foo {bar:.3{x}f} bar"`, the `foo `, `.3`, `f` and ` bar` are `FStringMiddle` tokens (4 in total).
>
> In the lambda example (`f"{lambda x: {x}"`) as there's a whitespace after `:` and before the `{` token, that is a `FStringMiddle` token whose content is a single space character. The lambda pattern in LALRPOP doesn't contain `FStringMiddle` token which means it'll error with an unexpected token. But, if you remove the space (`f"{lambda x:{x}}"`), there won't be a `FStringMiddle` token anywhere and it'll match the lambda pattern in LALRPOP definition.
>
> We'll emit an empty `FStringMiddle` token so that the parser will throw an error and disallow non-parenthesized lambda expressions. The way it is done is:
> 1. A `}` token while lexing for `FStringMiddle` is only encountered when we're inside a format spec (the part after the `:`).
> 2. We'll emit an empty `FStringMiddle` token when we encounter `}` and reduce the format spec depth to avoid going into infinite mode where the lexer then keeps emitting the empty token.
